docs/language/learn-ql/cpp/value-numbering-hash-cons.rst

```diff
@@ -0,0 +1,104 @@
+Hash consing and value numbering
+=================================================
+Overview
+--------
+In C and C++ QL databases, each node in the abstract syntax tree is represented by a separate object. This allows both analysis and results display to refer to specific appearances of a piece of syntax. However, it is frequently useful to determine whether two expressions are equivalent, either syntactically or semantically.
+
+The hash consing library (defined in ``semmle.code.cpp.valuenumbering.HashCons``) provides a mechanism for identifying expressions that have the same syntactic structure. The value numbering library (defined in ``semmle.code.cpp.ir.ValueNumber``) provides a mechanism for identifying expressions that compute the same value at runtime.
+
+
+Example C code
+--------------
+
+In the following C program, ``x + y`` and ``x + z`` will be assigned the same value number but different hash conses.
+
+.. code-block:: c
+
+    int x = 1;
+    int y = 2;
+    int z = y;
+    if(x + y == x + z) {
+      ...
+    }
+
+However, in the next example, the uses of ``x + y`` will have different value numbers but the same hash cons.
+
+.. code-block:: c
+
+    int x = 1;
+    int y = 2;
+    if(x + y) {
+      ...
+    }
+
+    x = 2;
+
+    if(x + y) {
+      ...
+    }
+
+Value numbering
+---------------
+The value numbering library (defined in ``semmle.code.cpp.valuenumbering.GlobalValueNumbering``) provides a mechanism for identifying expressions that compute the same value at runtime. Value numbering is useful when your primary concern is with the values being produced or the eventual machine code being run. For instance, value numbering might be used to determine whether a check is being done against the same value as the operation it is guarding.
```

Comment:
In the overview it says `semmle.code.cpp.ir.ValueNumber` and here it says `semmle.code.cpp.valuenumbering.GlobalValueNumbering`. Which is it?

Reply:
The latter - fixed in the overview.